compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2051,7 +2051,8 @@ private BType checkMappingConstructorCompatibility(BType bType, BLangRecordLiter
         if (tag == TypeTags.TYPEREFDESC) {
             BType refType = Types.getReferredType(bType);
             BType compatibleType = checkMappingConstructorCompatibility(refType, mappingConstructor);
-            return (refType.tag != TypeTags.UNION && refType.tag != TypeTags.INTERSECTION) ? bType : compatibleType;
+            return (!TypeTags.isUnionTypeTag(refType.tag) && refType.tag != TypeTags.INTERSECTION) ? bType :
```

Comment:
If the tag is not union/intersection, shouldn't we handle it as if we are handling non-union type without special-casing json etc ?

Reply:
As we are checking for the mapping constructor here, for the non-type referenced cases they are given as a map of the respective type.
 Eg. : 
 ```ballerina
anydata a = {"a": 1};
 ```
 if we have `anydata` then this method returns `map<anydata>` to construct the map. 

This is because we always expect a map type for constructing the map directly from runtime.

Reply:
Yes, I get that. There's anyway a bug in the logic. My question is if we perform the logic below this specific if check without returning; wouldn't it handle the mentioned cases?

Reply:
Blocked on https://github.com/ballerina-platform/ballerina-lang/issues/35237